Add request{Method,Mode,Destination} support.

yoshisatoyanagisawa opened the pull request:
The ServiceWorker static routing API has a condition on a request{Method,Mode,Destination}. This is a specification update to support that.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/6.html" title="Last updated on Jan 22, 2024, 10:26 AM UTC (b82af95)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/6/9709650...b82af95.html" title="Last updated on Jan 22, 2024, 10:26 AM UTC (b82af95)">Diff</a>

yoshisatoyanagisawa committed:
Add request{Method,Mode,Destination} support.

yoshisatoyanagisawa committed:
Add "." at the end of lines.

yoshisatoyanagisawa commented:
@domenic @sisidovski Can I ask you to review the spec update on request?

Also the question, but should I add RequestMode and RequestDestination in this specification's IDL?
Note that Chromium does not:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/service_worker/router_condition.idl

sisidovski commented on docs/index.bs:
```diff
@@ -1563,6 +1563,9 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
       dictionary RouterCondition {
         URLPatternCompatible urlPattern;
+        USVString requestMethod;
```
requestMethod in the fetch standard is defined as `ByteString` [1]. I wonder if it's fine to define this as USVString or not. 
[1] https://fetch.spec.whatwg.org/#request-class

domenic replied:
Matching the types from Fetch is best, so I agree that making it `ByteString` would be better.

This has slight observable consequences as it means that `requestMethod: "あ"` will throw immediately, instead of simply failing to match. A test for that might be good.

yoshisatoyanagisawa replied:
good catch!
I overlooked that.
I will also fix the Chromium code X(.

yoshisatoyanagisawa replied:
FYI: Chromium fix is https://chromium-review.googlesource.com/c/chromium/src/+/5223724

sisidovski reviewed (approved):
> Also the question, but should I add RequestMode and RequestDestination in this specification's IDL?

I believe it's fine not to define here.

domenic commented on docs/index.bs:
```diff
@@ -3271,6 +3277,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
               1. If running [=match=] with |pattern| and |request|'s [=request/URL=] returns null, [=continue=].
+          1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/requestMethod}}"] [=map/exists=], then:
+              1. Let |method| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/requestMethod}}"].
+              1. If |request|'s [=request/method] is not |method|, [=continue=].
```
```suggestion
              1. If |request|'s [=request/method=] is not |method|, [=continue=].
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3271,6 +3277,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. If |request|'s [=request/method] is not |method|, [=continue=].
+          1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/requestMode}}"] [=map/exists=], then:
+              1. Let |mode| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/requestMode}}"].
+              1. If |request|'s [=request/mode] is not |mode|, [=continue=].
```
```suggestion
              1. If |request|'s [=request/mode=] is not |mode|, [=continue=].
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3271,6 +3277,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. If |request|'s [=request/mode] is not |mode|, [=continue=].
+          1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/requestDestination}}"] [=map/exists=], then:
+              1. Let |destination| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/requestDestination}}"].
+              1. If |request|'s [=request/destination] is not |destination|, [=continue=].
```
```suggestion
              1. If |request|'s [=request/destination=] is not |destination|, [=continue=].
```

yoshisatoyanagisawa replied:
done.

domenic reviewed (commented):
>> Also the question, but should I add RequestMode and RequestDestination in this specification's IDL?
>
> I believe it's fine not to define here.

+1. They will automatically be cross-linked to their original definitions, as you can see from the preview: https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/6.html#installevent-interface

yoshisatoyanagisawa committed:
Address comments.

- Use ByteString instead USVString for requestMethod.
- typos on forgetting "=" in closing "=]".

yoshisatoyanagisawa commented:
> > > Also the question, but should I add RequestMode and RequestDestination in this specification's IDL?
> > 
> > 
> > I believe it's fine not to define here.
> 
> +1. They will automatically be cross-linked to their original definitions, as you can see from the preview: https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/6.html#installevent-interface

Thanks for clarifying and checking that.

yoshisatoyanagisawa commented:
Thanks for the review.  I am going to merge this.
